lib/internal/Magento/Framework/App/Utility/IPAddressNormaliser.php

```diff
@@ -0,0 +1,98 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+namespace Magento\Framework\App\Utility;
+
+use IPTools\Network;
+use IPTools\Range;
+
+/**
+ * Utility class to normalise list of IP addresses
+ */
+class IPAddressNormaliser
+{
+    /**
+     * Normalise / canonicalise list of IP addresses / ranges
+     *
+     * @param  string[] $addresses
+     * @return string[]
+     */
+    public function execute(array $addresses): array
+    {
+        $addresses = array_filter($addresses); // remove empty strings
+
+        $networks = [];
+        foreach ($addresses as $address) {
+            $networks[] = Network::parse(trim($address));
+        }
+
+        $networks = array_unique($networks); // remove obvious duplicates
+
+        do {
+            $startCount = count($networks);
+            $this->combineAdjacentNetworks($networks);
+            $this->removeOverlappingNetworks($networks);
+            $endCount = count($networks);
+        } while ($startCount != $endCount);
+
+        $networks = array_map(function ($network) {
+            return (string) $network;
+        }, $networks);
+
+        return $networks;
+    }
+
+    /**
+     * Combine adjacent networks
+     *
+     * @param string[] $networks
+     * @return void
+     */
+    private function combineAdjacentNetworks(array &$networks): void
+    {
+        if (count($networks) > 1) {
+            // Sort the list so we can compare each item with the following to
+            // determine if they are adjacent.
+            sort($networks, SORT_NATURAL);
+
+            // Define end point here as we expect the array to grow within the loop
+            $penultimate = count($networks) - 2;
+            for ($i = 0; $i <= $penultimate; $i++) {
+                $a = $networks[$i];
+                $b = $networks[$i + 1];
+                if ($a->getLastIP()->next() == $b->getFirstIP()) {
```

Comment:
can we use === here?

Reply:
We can; should we?

Reply:
Actually, no we can't. These functions both return an IP object. We could cast them to strings and strictly compare those, but I think the current implementation is more legible than that would be.

Reply:
Usually using == instead of === in some cases causing not obvious issues in not obvious cases. I really prefer to have strict comparison here, even if we'll add some string comparison

Reply:
I find this harder to read, but can commit if you'd like?
![Screenshot_2020-11-26_21-59-20](https://user-images.githubusercontent.com/334786/100393773-b8b06c80-3032-11eb-80f0-91555168eda9.png)

Reply:
yeah! somethins like that!